src/i18n.ts

```diff
@@ -0,0 +1,316 @@
+import {
+  Context,
+  Fluent,
+  type FluentBundleOptions,
+  FluentType,
+  type FluentValue,
+  type LocaleId,
+  type MiddlewareFn,
+  type NextFunction,
+  resolve,
+  type TranslationContext,
+} from "./deps.ts";
+
+import { readLocalesDir, readLocalesDirSync } from "./utils.ts";
+
+import type { I18nConfig, I18nFlavor, TranslateFunction } from "./types.ts";
+
+class FluentContext extends FluentType<string> {
+  constructor(value: string, private defaultValue = "") {
+    super(value);
+  }
+
+  // deno-lint-ignore no-explicit-any
+  toString(scope: any): string {
+    let ctx = JSON.parse(scope.args?.ctx as string);
+    const keys = this.value.split(".");
+    for (const key of keys) {
+      ctx = ctx[key];
+      if (ctx === undefined) break;
+    }
+    return ctx ?? this.defaultValue;
+  }
+}
+
+/**
+ * A custom Fluent function for accessing `Context` properties directly from
+ * Fluent translation files.
+ */
+export function CTX(args: FluentValue[]) {
```

Comment:
I don't really like having this funky function. We shouldn't encourage people to write any logic in templates

Reply:
I don't think it is logic, it just query the context properties directly instead export it to the translation context. Why do you think CTX can do the logic work?

Reply:
Not necessarily the logic. But why would you want to use context in the locale? Message text for example, what else? If so, why not create a new variable?

Reply:
That depends on how you plan to use it. You can use it to access to session variables You can use it to make fluent scripts more independent from source code. Because of fallback ability, you can always safely make sure that every scripts always return a nice text rather than some missing string because of the bot fail to get data or the bot don't support that function.


```FLT
greeting = hello {{ CTX('from.first_name', '?') }}
```

Reply:
I don't want my locales to be context-dependent.
I'd rather write 
```
greeting = hello {{ name }}
```
and pass name as `ctx.from.first_name ?? '?'`

Reply:
It's ok, we can vote for the feature, I don't mind at all. You can suggest this discussion to the telegram like you did we the above to see what other people think about it. (It's what I plan to do for the whole plugin anyway)

Reply:
@Loskir I see your point. I intially added the custom function to make the work easy. And as something better than current i18n's `from` and `chat` shortcuts. If more people disagree with having this, we can remove it. I just never got any negative feedback about this before.

> I don't want my locales to be context-dependent.

You don't have to! You don't have to make them context-dependent. There's no forcing to use it. You can always do it in the normal way.

The thing is that, the developer knows what the message needs as variables. If they're from context, its easy to just write them using CTX. They can also provide proper values for the values that can be undefined as well (it can also be done by specifying them in `t`, as we did before). But isn't

```ts
// id: "Your ID is: { CTX("from.id") }"
ctx.t("id")
```

much simpler and easier than this?:

```ts
// id: "Your ID is: {$id}"
ctx.t("id", { id: ctx.from.id })
```

Reply:
This function is not type-safe and is encouraging users to use hacky constructions and move business logic to locales. For example: what if ctx.from is not defined? They get no error about it in the first case but not in the second. 
And imagine the implementation being different in different locales!

Reply:
I agree with the error thing. But why should ctx.from be undefined? Doesn't the one who writes the code knows whats undefined and whats not undefined? Why should they use CTX("from.id") in their locales if ctx.from is not defined 😕 I could be wrong, can you provide a proper example?

Reply:
It's almost like using javascript instead of typescript. Yes, it works, as long as you keep everything in your head and make sure you don't make any mistakes. And since we are using typescript where we can almost guarantee the code is correct, why should we allow such an obvious step back?
If users know what they're doing, they can always add it themselves, it's not like we're sealing the API shut. But I don't think it should be added to the main library because of its bad influence on code quality.

Reply:
Hmm. Now I kinda agree with you :D I'll remove them in the next push. Thanks for making me think straight.

Reply:
> For example: what if ctx.from is not defined? They get an error about it in the second case but not in the first. 

That why we have the default arg for CTX, isn't it. And I personally think setting default value in the translated script will be better than we do that in code. The data belong to the translation, why should we keep a part of it in logic code?

My point is from a translator pov so I think CTX is a better choice. And if you ask me that which should be priority? I would say usability for translating. Because this is the i18n plugin, so it should be made for translators first.


After a long good discussion, I think your point also important... Maybe we can combine our idea into a new one instead of kill off a whole feature. I'll write more about this at the end of the day.

Reply:
I deleted previous command since I think fluent is powerful enough to handle all the necessary tasks but I still thinking about a fallback way. Which do you prefer:

The fluent way:
```FLT
-link = { $host }
visit = Visit { -link(host: "example.com") } for more information.
```

Or using `CTX` function:
```FLT
visit = Visit { -CTX("link", "example.com") } for more information.
```

If you think `CTX` is preferable then we can go with this. We still keep `CTX` but it's only allowed to access the declare context from `translate` function. That way we can still keep types pretty strict and offer a good solution on fallback variables for Fluent. How do you think?

Reply:
Essentially, having `CTX` is convenient, but it comes at the cost the we completely mess with the abstractions here. Handling incoming updates via a context object is clearly not related to injecting variables. Tying these two things together should be avoided, so  I agree with @Loskir and would prefer not to have `CTX`.

However, that does not mean that this isn't a valid use case. In practice, people will of course often use variables that are present on the context object, so it would be good to offer them something here. I think https://github.com/grammyjs/i18n/pull/20#discussion_r956710491 is a great idea for that.

Reply:
I also agree with the idea @ngdangtu-vn shared. Give me a two days, I'll implement that.

Reply:
@ngdangtu-vn 

1. I don’t think the two examples are equivalent, are they? The fluent way doesn’t take any external variables at all
 2. The main advantage of ctx function is that you can provide a default value right in the locales. You can’t pass undefined or null as fluent variables, so I don’t think we can work around it. Like for example (won’t work) `{ $name ?? "No Name" }`. We need to do research for a fluent way to manage it